contributors/design-proposals/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,649 @@
+has no resource stored in the API. Unlike `kubectl exec`, a Debug Container
+*does* have a `ContainerStatus` that is reported with `v1.PodStatus` and
+displayed by `kubectl describe pod`.
+
+For example, the following command would attach to a newly created container in
+a pod:
+
+```
+kubectl debug -it -c debug-shell --image=debian target-pod -- bash
+```
+
+Stdin and TTY default to true, and it would be reasonable for Kubernetes to
+provide a default container name and image, making the minimal possible debug
+command:
+
+```
+kubectl debug target-pod
+```
+
+This creates an interactive shell in a pod which can examine and signal all
+processes in the pod. It has access to the same network and IPC as processes in
+the pod. It can access the filesystem of other processes by `/proc/$PID/root`.
+As is already the case with regular containers, Debug Containers can enter
+arbitrary namespaces of another container via `nsenter` if run with
+`CAP_SYS_ADMIN`.
+
+*Please see the User Stories section for additional examples and Alternatives
+Considered for the considerable list of other solutions we considered.*
+
+## Implementation Details
+
+The implementation of `kubectl debug` closely mirrors the implementation of
+`kubectl exec`, with most of the complexity being implemented in the `kubelet`.
+From the perspective of the client, there's a new subresource of pod, /`debug`,
+that creates a Debug Container based on a provided `v1.Container` and attaches
+to its console. Users give Debug Containers a name (e.g. "debug" or "shell")
+which is used in subsequent interactions and is reported in `kubectl describe`.
+
+### Kubernetes API Changes
+
+Most of the complexity for Debug Containers is implemented in the kubelet, but
+the API server needs to proxy the request from the client to the kubelet. The
+most flexible approach is to allow the client to provide a `v1.Container` to the
+kubelet.
+
+We'll create a new API Object:
+
+```
+// PodDebugContainer describes a container to attach to a running pod for troubleshooting.
+type PodDebugContainer struct {
+        metav1.TypeMeta
+
+        // Template defines the pods that will be created from this pod template
+        Container v1.Container
+}
+```
+
+The pod API will gain a new `/debug` subresource that allows the following:
+
+1.  A `POST` of a `PodDebugContainer` to
+    `/api/v1/namespaces/$NS/pods/$POD_NAME/debug/$NAME` will create Debug
+    Container named `$NAME` running in pod `$POD_NAME`.
+1.  A `DELETE` of `/api/v1/namespaces/$NS/pods/$POD_NAME/debug/$NAME` will stop
+    the Debug Container `$NAME` in pod `$POD_NAME`.
+
+Once created, a client can attach to the console of a debug container using the
+existing attach endpoint, `/api/v1/namespaces/$NS/pods/$POD_NAME/attach`. Status
+of Debug Containers is reported via a new field in `v1.PodStatus`, described in
+a subsequent section.
+
+#### Alternative: Extending `/exec`
```

Comment:
@dchen1107 @lavalamp @smarterclayton @pwittrock 

For pod troubleshooting we need to choose between the object-based approach described above (suggested by @smarterclayton ) and the exec-based approach described below (suggested by @lavalamp and resembling the "image exec" approach of the original proposal, interestingly).